### PR TITLE
[21051] DynamicType to IDL serializer

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
## Description

This PR implements functionality to serialize a DynamicType to IDL.

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_ Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- _N/A_ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [x] Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 
https://github.com/eProsima/Fast-DDS-docs/pull/865
- _N/A_ Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/xtypes/utils.cpp [hunk 1]
```diff
@@ -0,0 +1,1477 @@
...
+}
+
+std::vector<uint32_t> container_size(
+        const DynamicType::_ref_type& dyn_type)
+{
+    TypeDescriptor::_ref_type type_descriptor {traits<TypeDescriptor>::make_shared()};
+    const auto ret = dyn_type->get_descriptor(type_descriptor);
+    if (ret != RETCODE_OK)
+    {
+        //throw utils::InconsistencyException("No Type Descriptor");
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/utils.cpp:1088]: And add log warnings.

File: src/cpp/fastdds/xtypes/utils.cpp [hunk 2]
```diff
@@ -0,0 +1,1477 @@
...
+std::string sequence_kind_to_str(
+        const DynamicType::_ref_type& dyn_type)
+{
+    auto internal_type = container_internal_type(dyn_type);
+    auto this_sequence_size = container_size(dyn_type);
+
+    std::stringstream ss;
+    ss << "sequence<" << type_kind_to_str(internal_type);
+
+    for (const auto& bound : this_sequence_size)
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/utils.cpp:1144]: Review this logic.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/utils.cpp:1144]: Compare bound with `static_cast<uint32_t>(LENGTH_UNLIMITED)` to cover the unbounded case.

**Comment by juanlofer-eprosima** [comment]
Consider extending conversions to take annotations into account.

**Comment by Tempate** [comment]
@richiprosima please test_3 this

File: test/unittest/dds/xtypes/type_conversion/types/idls/alias_struct.idl [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+typedef unsigned long MyLong;
+
+typedef MyShort MyRecursiveShort;
+
+struct alias_struct
+{
+    MyShort my_short;
+    MyLong my_long;
+    MyRecursiveShort my_recursive_short;
+};
```
Comment (richiware) [inline test/unittest/dds/xtypes/type_conversion/types/idls/alias_struct.idl:12]: We have a common repository containing the IDL files that are testes in all products ... or will be. Fast-DDS already has this repository in `thirdparty/dds-types-test`. I suggest to use these IDL files instead of new ones.

**Review by richiware — CHANGES_REQUESTED** [review]

File: test/unittest/dds/xtypes/type_conversion/types/idls/alias_struct.idl (hunk 3 above)
Comment (Tempate) [reply test/unittest/dds/xtypes/type_conversion/types/idls/alias_struct.idl:12]: That'd be ideal, yes. The problem is that the DynamicType to IDL serializer serializes a single DynamicType to IDL. The IDLs in `thirdparty/dds-types-test`, however, contain many different types. Furthermore, the DynamicType to IDL serializer can omit information present in the IDL but not present in the DynamicType. For example, a DynamicType has no way of knowing if value was explicitly set by the user to its default value, or if the user didn't set it. Any ideas on how to fix these issues?

Comment (richiware) [reply test/unittest/dds/xtypes/type_conversion/types/idls/alias_struct.idl:12]: Maybe in the future, when we have the IDL runtime parser we can test DynamicType -> IDL -> DynamicType and compare the original DynamicType and the result.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:1]: Perhaps it is better to have a linked cpp file instead.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:1]: And to make it consistent with https://github.com/eProsima/Fast-DDS/pull/4780 , please use the same folder structure and rename to `dynamic_type_idl.cpp`.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:1]: Also, remember to add the new sources to the corresponding CMakeLists.txt files.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 4]
```diff
@@ -0,0 +1,1116 @@
...
+    }
+
+    return ret;
+}
+
+ReturnCode_t map_kind_to_str(
+        const DynamicType::_ref_type& dyn_type,
+        std::string& map_str) noexcept
+{
+    if (dyn_type->get_kind() != TK_MAP)
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:351]: I'd say this should better be an assert.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 5]
```diff
@@ -0,0 +1,1116 @@
...
+        const DynamicType::_ref_type& dyn_type,
+        std::vector<std::pair<std::string, MemberDescriptor::_ref_type>>& result) noexcept
+{
+    ReturnCode_t ret = RETCODE_OK;
+
+    // Skip the inherited members
+    TypeDescriptor::_ref_type type_descriptor {traits<TypeDescriptor>::make_shared()};
+    ret = dyn_type->get_descriptor(type_descriptor);
+
+    if (ret != RETCODE_OK)
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:456]: Perhaps there should be more logs when something goes wrong to give hints to the user.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 6]
```diff
@@ -0,0 +1,1116 @@
...
+    return ret;
+}
+
+ReturnCode_t array_kind_to_str(
+        const DynamicType::_ref_type& dyn_type,
+        std::string& array_str) noexcept
+{
+    if (dyn_type->get_kind() != TK_ARRAY)
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Type is not an array.");
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:308]: I've been using warnings in the DynamicData->JSON part, not sure which one is better, but I guess we should be using the same kind.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 7]
```diff
@@ -0,0 +1,1116 @@
...
+        return ret;
+    }
+
+    const auto base_type = type_descriptor->base_type();
+
+    std::uint32_t first_member = 0;
+
+    if (base_type != nullptr)
+    {
+        // If the struct has a base type, the first member is the base type
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:467]: Please rephrase, I understand what you are doing but I find the sentence misleading.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:467]: And add a note/warning saying that it is assumed this is the case (base members are always before derived ones).

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 8]
```diff
@@ -0,0 +1,1116 @@
...
+            return ret;
+        }
+
+        MemberDescriptor::_ref_type member_descriptor {traits<MemberDescriptor>::make_shared()};
+        ret = member->get_descriptor(member_descriptor);
+
+        if (ret != RETCODE_OK)
+        {
+            return ret;
+        }
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:488]: You could benefit from the Impl class to simplify the code a bit, as well as for efficiency.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 9]
```diff
@@ -0,0 +1,1116 @@
...
+            sequence_str += ", " + std::to_string(bound);
+        }
+    }
+
+    sequence_str += ">";
+
+    return ret;
+}
+
+ReturnCode_t get_members_sorted(
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:446]: If this method is designed to only work with structs, you may add an assert and/or make it explicit in the method name.

File: include/fastdds/dds/xtypes/utils.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+#include <string>
+
+#include <fastdds/dds/xtypes/dynamic_types/DynamicType.hpp>
+
+namespace eprosima {
+namespace fastdds {
+namespace dds {
+
+/**
+ * @brief Serializes a DynamicType into an IDL schema.
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/dds/xtypes/utils.hpp:31]: IMO IDL representation is best

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.hpp [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*!
+ * @file dyn_type_tree.hpp
+ */
+
+#ifndef _FASTDDS_DDS_XTYPES_DYN_TYPE_TREE_HPP_
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.hpp:19]: It should be consistent since #4958, at least for the public headers (the ones under `include`). We ought to follow that convention for all new headers (private or otherwise).

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 5 above)
Comment (EduPonz) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:456]: One addition, on the comparisons, we prefer to place the constant on the LHS

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 12]
```diff
@@ -0,0 +1,1116 @@
...
+    {
+        const auto UNBOUNDED = static_cast<std::uint32_t>(LENGTH_UNLIMITED);
+
+        if (bound != UNBOUNDED)
+        {
+            sequence_str += ", " + std::to_string(bound);
+        }
+    }
+
+    sequence_str += ">";
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:441]: If you are keeping only the ostream overload, perhaps it'd be better to use << everywhere. I believe it's also more efficient but I'm not sure.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 13]
```diff
@@ -0,0 +1,1116 @@
...
+
+    BoundSeq bounds;
+    ret = container_size(dyn_type, bounds);
+
+    if (ret != RETCODE_OK)
+    {
+        return ret;
+    }
+
+    for (const auto& bound : bounds)
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:431]: I don't think it's possible to have multiple dimensions in sequences.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:431]: Also, I find this logic a bit odd, in the sense that I believe the only case where UNBOUNDED may be present it's in the first (unique) dimension.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 14]
```diff
@@ -0,0 +1,1116 @@
...
+    }
+
+    if (type_descriptor->bound().size() != 1)
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Bitmask type must have exactly one bound.");
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    // Annotation with the bitmask size
+    static constexpr std::uint32_t DEFAULT_BITMASK_SIZE = 32;
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:688]: Can't you access the default value instead of hardcoding it?

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 6 above)
Comment (Tempate) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:308]: I believe errors are more pertinent.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 15]
```diff
@@ -0,0 +1,1116 @@
...
+    // Annotation with the extensibility kind
+    TypeDescriptor::_ref_type type_descriptor {traits<TypeDescriptor>::make_shared()};
+    ret = node.info.dynamic_type->get_descriptor(type_descriptor);
+
+    if (ret != RETCODE_OK)
+    {
+        return ret;
+    }
+
+    switch (type_descriptor->extensibility_kind())
```
[the remark this answers is not in the document]
Comment (Tempate) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:900]: It's written in the documentation. In theory, excluding some annotations, everything is supported.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 7 above)
Comment (Tempate) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:467]: I have rephrased the sentence, but I don't think a note or warning are necessary since it's part of the standard. In Chat-GPT's words: "According to the xtypes specification, when you derive one type from another, the base class members must indeed have lower indices than the members of the derived class. "

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 9 above)
Comment (Tempate) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:446]: I have removed the method all together. It was only being used once and it repeated calculations.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 14 above)
Comment (Tempate) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp:688]: The default value is not accessible.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 6 above)
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp]: Ok, changed warnings for errors on my part.

File: src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp [hunk 16]
```diff
@@ -0,0 +1,1086 @@
...
+    {
+        return ret;
+    }
+
+    internal_type = type_descriptor->element_type();
+
+    return ret;
+}
+
+ReturnCode_t container_size(
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp:535]: NIT: since this now also applies to other types other than containers you might rename it to `get_bounds` or similar.

File: src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp [hunk 17]
```diff
@@ -0,0 +1,1086 @@
...
+    ret = dyn_type->get_descriptor(type_descriptor);
+
+    if (RETCODE_OK != ret)
+    {
+        return ret;
+    }
+
+    bounds = type_descriptor->bound();
+
+    static const auto UNBOUNDED = static_cast<std::uint32_t>(LENGTH_UNLIMITED);
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp:551]: Could you try to make this constexpr?

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 18]
```diff
@@ -0,0 +1,1116 @@
...
+{
+    ReturnCode_t ret = RETCODE_OK;
+
+    switch (type->get_kind())
+    {
+        case TK_STRUCTURE:
+        {
+            // If is struct, the call is recursive.
+            // Create new tree node
+            utilities::collections::TreeNode<TreeNodeType> parent(member_name, type->get_name().to_string(), type);
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp]: I think you can set `using TreeNode = utilities::collections::TreeNode<TreeNodeType>` , in case you prefer that option.

File: test/unittest/dds/xtypes/type_conversion/types/idls/inheritance_struct.idl [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+struct FatherStruct : GrandfatherStruct
+{
+    long father;
+};
+
+struct SonStruct : FatherStruct
+{
+    long son;
+};
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply test/unittest/dds/xtypes/type_conversion/types/idls/inheritance_struct.idl]: I was more referring to having for example:

```
struct ParentStruct : GrandparentStruct
{
    short my_short;
    string my_string;
    GrandparentStruct my_grand_parent_father;
};
```

and/or

```
struct GrandparentStruct
{
    long my_long;
    ParentStruct my_parent_son;
};
```

[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply test/unittest/dds/xtypes/type_conversion/types/idls/inheritance_struct.idl]: Not relevant anyway.

File: test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+    SupportedTypes::STRUCT_STRUCT,
+    SupportedTypes::UNION_STRUCT
+};
+
+void register_type_object_representation(
+        const std::string& type_name)
+{
+    using namespace eprosima::fastdds::dds;
+
+    if (type_name == SupportedTypes::ALIAS_STRUCT)
```
Comment (juanlofer-eprosima) [inline test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp:103]: This must hurt for a switch-lover. You could maybe use an enumeration and define a to_string method. Ignore this if you will.

File: test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    {
+        TypeSupport type_alias_struct(new AliasStructPubSubType());
+        type_alias_struct->register_type_object_representation();
```
Comment (juanlofer-eprosima) [inline test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp:106]: Can't you move this outside from the conditions?

Comment (juanlofer-eprosima) [reply test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp:106]: Way better now BTW :+1:

File: versions.md [hunk 22]
```diff
@@ -68,6 +68,7 @@ Forthcoming
 * New `DomainParticipantExtendedQos` that includes both `DomainId` and `DomainParticipantQos` (extends `DomainParticipantFactory` API).
 * Make Blackbox tests not include any private API.
 * Remove all the private API include from Blackbox tests.
+* New `idl_serialize` method to convert a `DynamicType` to an IDL string.
```
Comment (juanlofer-eprosima) [inline versions.md:71]: NIT: maybe rephrase as now it isn't really an string.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp (hunk 15 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp]: I leave it unresolved to double check before merging.

File: src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp [hunk 23]
```diff
@@ -0,0 +1,1086 @@
...
+            TreeNode<TreeNodeType> parent(member_name, type->get_name().to_string(), type);
+
+            // Get its base class
+            TypeDescriptor::_ref_type type_descriptor {traits<TypeDescriptor>::make_shared()};
+            ret = type->get_descriptor(type_descriptor);
+
+            if (RETCODE_OK != ret)
+            {
+                return ret;
+            }
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp:67]: I'm still not satisfied with the amount of traces present for failing scenarios. Not sure to what extent this is relevant. I insist in suggesting to cast to the Impl class, you would be avoiding a big amount of return codes to check.

File: test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp (hunk 20 above)
Comment (Tempate) [reply test/unittest/dds/xtypes/serializers/idl/types/all_types.hpp:103]: I had that at the beginning. It's the most logical approach. The problem was that I wasn't able to display with GoogleTest the `to_string`  name of the tests, so I kept getting:
```
The following tests passed:
        DynTypeIDLTests/DynTypeIDLTests.to_idl/0
        DynTypeIDLTests/DynTypeIDLTests.to_idl/1
        ...
```
instead of
```
The following tests passed:
        DynTypeIDLTests/DynTypeIDLTests.to_idl/"alias_struct"
        DynTypeIDLTests/DynTypeIDLTests.to_idl/"array_struct"
        ...
```

File: src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp [hunk 24]
```diff
@@ -0,0 +1,1086 @@
...
+
+    if (RETCODE_OK != ret)
+    {
+        return ret;
+    }
+
+    idl << "typedef ";
+
+    // Find the base type of the alias
+    ret = type_kind_to_idl(type_descriptor->base_type(), idl);
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/xtypes/serializers/idl/dynamic_type_idl.cpp:662]: Be careful here, we must make sure typedef dependencies are written in the right order to produce a valid IDL.

File: src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp [hunk 25]
```diff
@@ -0,0 +1,1116 @@
...
+            {
+                return ret;
+            }
+
+            container.add_branch(node);
+
+            node = container;
+            break;
+        }
+        default:
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/type_conversion/dyn_type_tree.ipp]: And probably also with aliases.

**Comment by juanlofer-eprosima** [comment]
Add the new IDL files to `update_generated_code_from_idl.sh`.

**Comment by juanlofer-eprosima** [comment]
> Add the new IDL files to `update_generated_code_from_idl.sh`.

By appending the following to `files_needing_output_dir`

'./test/unittest/dds/xtypes/serializers/idl/types/alias_struct/alias_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/array_struct/array_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/bitmask_struct/bitmask_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/bitset_struct/bitset_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/enum_struct/enum_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/extensibility_struct/extensibility_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/key_struct/key_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/map_struct/map_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/primitives_struct/primitives_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/sequence_struct/sequence_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/string_struct/string_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/struct_struct/struct_struct.idl|./gen'
  './test/unittest/dds/xtypes/serializers/idl/types/union_struct/union_struct.idl|./gen'

File: include/fastdds/dds/xtypes/utils.hpp [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+{
+    OMG,
+    EPROSIMA,
+};
+
+/**
+ * @brief Serializes a @ref DynamicType into its IDL representation.
+ *
+ * @param [in] dynamic_type The @ref DynamicType to serialize.
+ * @param [in,out] output The IDL representation.
```
[the remark this answers is not in the document]
Comment (Tempate) [reply include/fastdds/dds/xtypes/utils.hpp:43]: Your comments are always enigmatic. I don't know what you're talking about.

File: src/cpp/fastdds/xtypes/utils.cpp [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+        const DynamicType::_ref_type& dynamic_type,
+        std::ostream& output) noexcept
+{
+    // Create a tree representation of the dynamic type
+    utilities::collections::TreeNode<TreeNodeType> parent_type;
+    const auto ret = dyn_type_to_tree(dynamic_type, "PARENT", parent_type);
+
+    if (ret != RETCODE_OK)
+    {
+        return ret;
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/xtypes/utils.cpp:49]: But log error right?

File: include/fastdds/dds/xtypes/utils.hpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply include/fastdds/dds/xtypes/utils.hpp:44]: I meant that I don't find the sentence very accurate, sounds to me like "output === The IDL representation", instead of something like "\c std::ostream reference where the IDL representation is inserted/streamed."

File: test/unittest/dds/xtypes/serializers/idl/CMakeLists.txt [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+    ${CMAKE_CURRENT_BINARY_DIR}/types/string_struct/string_struct.idl
+    COPYONLY)
+configure_file(
+    ${CMAKE_CURRENT_SOURCE_DIR}/types/struct_struct/struct_struct.idl
+    ${CMAKE_CURRENT_BINARY_DIR}/types/struct_struct/struct_struct.idl
+    COPYONLY)
+configure_file(
+    ${CMAKE_CURRENT_SOURCE_DIR}/types/union_struct/union_struct.idl
+    ${CMAKE_CURRENT_BINARY_DIR}/types/union_struct/union_struct.idl
+    COPYONLY)
```
Comment (juanlofer-eprosima) [inline test/unittest/dds/xtypes/serializers/idl/CMakeLists.txt:66]:
```suggestion
file(GLOB_RECURSE IDL_FILES ${CMAKE_CURRENT_SOURCE_DIR}/types/**/*.idl)

foreach(idl_file ${IDL_FILES})
    file(RELATIVE_PATH relative_path ${CMAKE_CURRENT_SOURCE_DIR} ${idl_file})
    get_filename_component(relative_dir ${relative_path} DIRECTORY)
    file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${relative_dir})
    file(COPY ${idl_file} DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/${relative_dir})
endforeach()
```

Comment (juanlofer-eprosima) [reply test/unittest/dds/xtypes/serializers/idl/CMakeLists.txt:66]: Could you please also modify `test/unittest/dds/xtypes/serializers/json/CMakeLists.txt` with the following?

```
file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/types/comprehensive_type/json
    DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/types/comprehensive_type
    )
```

File: test/unittest/dds/xtypes/serializers/idl/types/alias_struct/alias_struct.idl [hunk 29]
```diff
@@ -0,0 +1,11 @@
...
+
+typedef short MyShort;
+
+typedef MyShort MyRecursiveShort;
+
+struct AliasStruct
+{
+    MyLong my_long;
+    MyRecursiveShort my_recursive_short;
+};
```
Comment (juanlofer-eprosima) [inline test/unittest/dds/xtypes/serializers/idl/types/alias_struct/alias_struct.idl:11]:
```suggestion
typedef unsigned long MyLong;

typedef short MyShort;

typedef MyShort MyRecursiveShort;

typedef int MyInt;

typedef MyInt MyRecursiveInt;

struct AliasStruct
{
    MyLong my_long;
    MyRecursiveShort my_recursive_short;
    MyRecursiveInt my_recursive_int;
    MyInt my_int;
};
```

Comment (juanlofer-eprosima) [reply test/unittest/dds/xtypes/serializers/idl/types/alias_struct/alias_struct.idl:11]: To cover also the case of inverse dependency order in members.

File: utils/scripts/update_generated_code_from_idl.sh [hunk 30]
```diff
@@ -45,6 +45,19 @@ files_needing_output_dir=(
     './thirdparty/dds-types-test/IDL/strings.idl|../../../test/dds-types-test'
     './thirdparty/dds-types-test/IDL/structures.idl|../../../test/dds-types-test'
     './thirdparty/dds-types-test/IDL/unions.idl|../../../test/dds-types-test'
+    './test/unittest/dds/xtypes/serializers/idl/types/alias_struct/alias_struct.idl|./gen'
```
Comment (juanlofer-eprosima) [inline utils/scripts/update_generated_code_from_idl.sh:48]: Sort

**Review by Tempate — DISMISSED** [review]
LGTM

**Review by richiware — APPROVED** [review]

**Review by Tempate — APPROVED** [review]
LGTM
